Changes for 3.0

carstenbauer opened the pull request:
Closes #47 

**Copied from there:**
A summary of the new API / changes that need to be made (likely incomplete and maybe some points are still up for debate):

* [x] new function `chunk(x; ...)` that returns an iterator that delivers the chunks, i.e. views into `x`. (We choose `chunk`, singular, here to indicate the verb form `to chunk`. This is to be in line with `enumerate`, `zip`, and co.)
* [x] rename `chunks(x; ...)` to `chunk_indices(x; ...)` and make the latter return an iterator that delivers the indices (of `x`) for each chunk.
* [x] ~~iterator names: `Chunk` (returned by `chunk`) and `ChunkIndices` (returned by `chunk_indices`). This is in line with `Base.Iterators.Enumerate` (returned by `enumerate`), `Base.Pairs` (returned by `pairs`), and `Base.Iterators.Zip` (returned by `zip`)~~ I implemented a single iterator `ChunksIterator` that has a type parameter `ReturnIndices` or `ReturnViews` to indicate what will be eventually returned.
* [x] drop the legacy API that relies on positional arguments
* [x] drop the `Symbol` option for `split`. While it is slightly more convenient, it can be less efficient and it doesn't make much sense to have two ways to do the same thing.
* [x] make `getchunk` internal (not public API anymore) + rename it to `getchunkindices`

What I think might make sense but I'm unsure about / not sure it's worth it:

* Should we rename `BatchSplit` to `ContiguousSplit`? The latter seems to communicate the critical point more clearly: the resulting chunks hold contiguous data.
* I don't like that we use "chunk" on the one hand and "split" on the other hand. Why two terms for the same thing? Perhaps something like `chunk(x; n=4, strategy=Contiguous())` or something less verbose makes more sense as it gets rid of "split" terminology entirely?

carstenbauer commented:
```julia-repl
julia> using ChunkSplitters

julia> function f(x)
           s = zero(eltype(x))
           for xc in chunk(x; n=4)
               s += sum(xc)
           end
           return s
       end
f (generic function with 1 method)

julia> x=rand(10^3);

julia> f(x) ≈ sum(x)
true

julia> function f_indices(x)
           s = zero(eltype(x))
           for idcs in chunk_indices(x; n=4)
               s += sum(@view(x[idcs]))
           end
           return s
       end
f_indices (generic function with 1 method)

julia> f_indices(x) ≈ sum(x)
true

julia> @benchmark f($x)
BenchmarkTools.Trial: 10000 samples with 788 evaluations.
 Range (min … max):  160.057 ns … 195.801 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     160.426 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   160.774 ns ±   1.690 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  ▄██▄                                                          ▂
  ████▇▆▅▄▅▅▅▆▇██▆▅▅▄▄▅▅▄▃▅▅▆▄▄▅▄▅▄▄▅▆▆▇▇▇███▇▇▅▆▅▅▅▅▃▅▁▄▄▄▄▁▄▆ █
  160 ns        Histogram: log(frequency) by time        170 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @benchmark f_indices($x)
BenchmarkTools.Trial: 10000 samples with 788 evaluations.
 Range (min … max):  159.688 ns … 181.313 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     159.898 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   160.350 ns ±   1.865 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  ▇█▄▂      ▁                                                   ▁
  █████▆▆▆▆▇█▇▇▅▆▅▄▄▄▁▁▆▅▄▅▅▄▃▃▅▅▇▆▆▇██▇▆▅▆▆▅▄▃▅▄▄▅▅▅▄▅▆▆▃▄▆▃▄▇ █
  160 ns        Histogram: log(frequency) by time        171 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.
```

carstenbauer commented:
Revamped documentation preview: https://juliafolds2.github.io/ChunkSplitters.jl/previews/PR48/

carstenbauer commented:
On 1.10, `chunk(x; size=10)` seems to allocate... Why?

```julia
function f_chunk(x; n=nothing, size=nothing)
    s = zero(eltype(x))
    for xdata in chunk(x; n=n, size=size)
        s += sum(xdata)
    end
    return s
end
x = rand(10^3)
b = @benchmark $f_chunk($x; n=4) samples = 1 evals = 1
@test b.allocs == 0
b = @benchmark $f_chunk($x; size=10) samples = 1 evals = 1
if VERSION < v"1.11"
    @test_broken b.allocs == 0 # TODO: why does this allocate on 1.10?
else
    @test b.allocs == 0
end
```

```julia-repl
julia> b = @benchmark $f_chunk($x; size=10) samples = 1 evals = 1
BenchmarkTools.Trial: 1 sample with 1 evaluation.
 Single result which took 1.542 μs (0.00% GC) to evaluate,
 with a memory estimate of 6.19 KiB, over 99 allocations.
```

Note that `chunk_indices(x; size=10)` does not allocate.

carstenbauer commented:
I noticed that the svg in the old/current docs is wrong. Which one of the two do you prefer?

Indicating the scattering by coloring chunks:

![splitters_colors](https://github.com/user-attachments/assets/8e41c61c-af50-4ec4-b500-15c708c92209)

Indicating the scattering by arrows:
![splitters_arrows](https://github.com/user-attachments/assets/e0523fa3-9ded-4766-973c-503ea36abd64)

I think I prefer the latter. However, the former is kind of the package logo :D

carstenbauer commented:
Alright guys, as it turned out, I had the motivation to implement this right away. Please checkout the [new (preview) docs](https://juliafolds2.github.io/ChunkSplitters.jl/previews/PR48/) and run some tests to see if things are stable.

What I did beyond the points mentioned in the OP is drop 1.6 support (1.10 is the new lower bound).

Also, if you have an idea about https://github.com/JuliaFolds2/ChunkSplitters.jl/pull/48#issuecomment-2363287934 that would be great as well.

carstenbauer commented:
The natural next step is to test OhMyThreads.jl with this.

carstenbauer commented:
* [ ] add a transition guide, that highlights the most important breaking changes

MasonProtter commented:
I'll try to look at this tonight. I suspect there's an inference problem here on 1.10, but I'll dig in and see if we can figure out what went wrong.

lmiq commented:
> * Should we rename `BatchSplit` to `ContiguousSplit`? The latter seems to communicate the critical point more clearly: the resulting chunks hold contiguous data.
> 
> * I don't like that we use "chunk" on the one hand and "split" on the other hand. Why two terms for the same thing? Perhaps something like `chunk(x; n=4, strategy=Contiguous())` or something less verbose makes more sense as it gets rid of "split" terminology entirely?

I also like these changes, in any case.

Concerning https://github.com/JuliaFolds2/ChunkSplitters.jl/pull/48#issuecomment-2363287934 , I've tried now many things, including manually defining functions for each case instead of relying on dispatch for `_getchunkindices`, but nothing worked. I couldn't detect any type inference problem either.

carstenbauer commented:
> I also like these changes, in any case.

I still think that it might make sense to consider renaming `Batch → Consecutive` and `Scatter → RoundRobin`.

As for renaming `split` to `strategy` (or something similar), there are two reasons I didn't do it right away (but I'm still open to be convinced):
1) `strategy` feels long compared to `split`
2) The package name `ChunkSplitters.jl` already establishes "chunk" and "split" as terminology.

fredrikekre commented on src/internals.jl:
```diff
@@ -0,0 +1,200 @@
+struct FixedSize <: Constraint end
+
+abstract type ReturnType end
+struct ReturnIndices <: ReturnType end
+struct ReturnViews <: ReturnType end
```
> I implemented a single iterator ChunksIterator that has a type parameter ReturnIndices or ReturnViews to indicate what will be eventually returned.

A bit of a drive-by comment, but unless you expect the number of `ReturnType`s to expand, isn't it cleaner to have two types called e.g. `IndexChunks` and `Chunks` (or `DataChunks` or `ViewChunks`)?

carstenbauer replied:
Keeping them as one type saved me a little bit of code duplication. But, yeah, I think it's cleaner and worth it. (Done)

fredrikekre commented on src/internals.jl:
```diff
@@ -0,0 +1,200 @@
+end
+
+# public API
+function chunk(collection;
```
Why not `chunks`? It returns multiple `chunk`s. Or is this a verb and not a noun? To me `chunks(...)::Chunks` and `index_chunks(...)::IndexChunks` would feel more natural.

carstenbauer replied:
> Why not chunks? It returns multiple chunks. Or is this a verb and not a noun? To me chunks(...)::Chunks and index_chunks(...)::IndexChunks would feel more natural.

It's supposed to be a verb similar to `enumerate`, `partition`, and `zip`. However, we also have `pairs` and `eachindex`, so we could also use a noun instead. Frankly, I don't have a super strong opinion here. It has crossed my mind that `chunk` might be more likely to be interpreted as a noun and might lead to confusion.

lmiq replied:
I find more natural "chunks", but since I don't use this word in any other context in life I can get used to any alternative.

carstenbauer replied:
I went with @fredrikekre's suggestion.

fredrikekre reviewed (commented):
> One remaining minor question is whether we should export the subtypes <:Split

I prefer not exporting, in particular since the practice of always explicitly `using` symbols anyway is getting more common (https://github.com/ericphanson/ExplicitImports.jl).


> On the one hand, it is nice that one can just do using ChunkSplitters and write chunks(x; n=10, split=RoundRobin()).

This export is actually slightly annoying because the perfect variable name for the return value is `chunks`, and obviously `chunks = chunks(...)` doesn't work.

With the current export situation I would simply use
```
using ChunkSplitters: ChunkSplitters
```
and always use the `ChunkSplitters.` prefix for clarity.

carstenbauer commented on src/internals.jl:
```diff
@@ -0,0 +1,212 @@
+    size::Int
+end
+
+struct IndexChunksIterator{T,C<:Constraint,S<:Split} <: AbstractChunksIterator{T,C,S}
```
Bikeshedding indeed 😉 I'm fine dropping the `Iterator` suffix. Let me just note that

1) the primary use case is to iterate over these, and
2) this isn't public API anyways (so we can rename these types whenever we want)

lmiq replied:
I prefer exporting the symbols, not bringing them to the namespace is always an option for the user.

carstenbauer replied:
@lmiq, wrong thread? I assume by "the symbols" you mean `Consecutive` and `RoundRobin` and aren't suggesting to make the "iterator" types public API?

carstenbauer replied:
FWIW, I've dropped the "Iterator" suffixes.

carstenbauer commented:
> I prefer not exporting, in particular since the practice of always explicitly `using` symbols anyway is getting more common ([ericphanson/ExplicitImports.jl](https://github.com/ericphanson/ExplicitImports.jl)).
> 
> > On the one hand, it is nice that one can just do using ChunkSplitters and write chunks(x; n=10, split=RoundRobin()).

I like the trend of explicitly `using` the bits that one actually needs. However, for this case, it doesn't matter whether we export the `<:Split` types or not. The question is about what should be "imported" if one simply does `using ChunkSplitters`. Not exporting anything (are you suggesting that?) seems radical to me. The naive `using ChunkSplitters` is useful in interactive workflows

> This export is actually slightly annoying because the perfect variable name for the return value is `chunks`, and obviously `chunks = chunks(...)` doesn't work.
> 
> With the current export situation I would simply use
> 
> ```
> using ChunkSplitters: ChunkSplitters
> ```
> 
> and always use the `ChunkSplitters.` prefix for clarity.

Once again, not exporting `chunks`, the main API of this package, seems bad/too radical to me. And while I see your point about the `chunks` name clash, the same could be said about `size = size(...)`, `length = length(...)`, and probably many more functions. Moreover, the typical use case of `chunks(...)` is in combination with `for` or `map` in which case you don't store it in a variable anyways.

carstenbauer commented on src/internals.jl:
```diff
@@ -0,0 +1,212 @@
+struct Enumerate{I<:AbstractChunksIterator}
+    itr::I
+end
+enumerate(c::AbstractChunksIterator) = Enumerate(c)
```
I agree, I just kept the style that was there before this PR. I'm fine changing this.

lmiq replied:
I am fine with both alternatives, and we should probably follow some style guideline, if there is one.

fredrikekre commented:
Yea I just meant that if anything `chunks` would be the "problematic" export, not the split types.

lmiq commented:
Just to say that I think this PR is fine, but we need to find a solution to https://github.com/JuliaFolds2/ChunkSplitters.jl/pull/48#issuecomment-2363287934 .

fredrikekre commented:
https://github.com/JuliaFolds2/ChunkSplitters.jl/pull/50 fixes https://github.com/JuliaFolds2/ChunkSplitters.jl/pull/48#issuecomment-2363287934

carstenbauer commented:
I think this is good to go (unless we want to make other breaking changes).

lmiq commented:
Thanks a lot @carstenbauer, and others of course, great work!
